Using ListenLimit and TCPKeepAlive panics

```go
// main.go

package main

import (
	"log"
	"net/http"
	"time"

	"gopkg.in/tylerb/graceful.v1"
)

func main() {
	srv := &graceful.Server{
		ListenLimit:  10,
		TCPKeepAlive: 1 * time.Minute,
		Server: &http.Server{
			Handler: http.NewServeMux(),
			Addr:    ":8080",
		},
	}
	log.Println(srv.ListenAndServe())
}
```

```go
$ go run main.go
panic: interface conversion: net.Listener is *netutil.limitListener, not *net.TCPListener

goroutine 1 [running]:
panic(0x3184e0, 0xc82000e500)
	/usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:481 +0x3e6
github.com/djui/mb-go/vendor/gopkg.in/tylerb/graceful%2ev1.(*Server).Serve(0xc820080780, 0x1245640, 0xc8200b6760, 0x0, 0x0)
	/Users/abc/dev/go/src/github.com/djui/mb-go/vendor/gopkg.in/tylerb/graceful.v1/graceful.go:242 +0x81f
github.com/djui/mb-go/vendor/gopkg.in/tylerb/graceful%2ev1.(*Server).ListenAndServe(0xc820080780, 0x0, 0x0)
	/Users/abc/dev/go/src/github.com/djui/mb-go/vendor/gopkg.in/tylerb/graceful.v1/graceful.go:151 +0xdf
main.main()
	/Users/abc/dev/go/src/github.com/djui/mb-go/foo.go:20 +0x1a7
exit status 2
```

--- a/graceful.go
+++ b/graceful.go
@@ -237,7 +237,7 @@ func (srv *Server) Serve(listener net.Listener) error {
 	}
 
 	if srv.TCPKeepAlive != 0 {
-		listener = tcpKeepAliveListener{listener.(*net.TCPListener), srv.TCPKeepAlive}
+		listener = keepAliveListener{listener, srv.TCPKeepAlive}
 	}
 
 	// Make our stopchan
@@ -438,21 +438,28 @@ func (srv *Server) shutdown(shutdown chan chan struct{}, kill chan struct{}) {
 	srv.chanLock.Unlock()
 }
 
-// tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
+type keepAliveConn interface {
+	SetKeepAlive(bool) error
+	SetKeepAlivePeriod(d time.Duration) error
+}
+
+// keepAliveListener sets TCP keep-alive timeouts on accepted
 // connections. It's used by ListenAndServe and ListenAndServeTLS so
 // dead TCP connections (e.g. closing laptop mid-download) eventually
 // go away.
-type tcpKeepAliveListener struct {
-	*net.TCPListener
+type keepAliveListener struct {
+	net.Listener
 	keepAlivePeriod time.Duration
 }
 
-func (ln tcpKeepAliveListener) Accept() (c net.Conn, err error) {
-	tc, err := ln.AcceptTCP()
+func (ln keepAliveListener) Accept() (net.Conn, error) {
+	c, err := ln.Listener.Accept()
 	if err != nil {
-		return
+		return nil, err
 	}
-	tc.SetKeepAlive(true)
-	tc.SetKeepAlivePeriod(ln.keepAlivePeriod)
-	return tc, nil
+
+	kac := c.(keepAliveConn)
+	kac.SetKeepAlive(true)
+	kac.SetKeepAlivePeriod(ln.keepAlivePeriod)
+	return c, nil
 }

--- a/graceful.go
+++ b/graceful.go
@@ -10,8 +10,6 @@ import (
 	"sync"
 	"syscall"
 	"time"
-
-	"golang.org/x/net/netutil"
 )
 
 // Server wraps an http.Server with graceful connection handling.
@@ -235,7 +233,7 @@ func Serve(server *http.Server, l net.Listener, timeout time.Duration) error {
 func (srv *Server) Serve(listener net.Listener) error {
 
 	if srv.ListenLimit != 0 {
-		listener = netutil.LimitListener(listener, srv.ListenLimit)
+		listener = LimitListener(listener, srv.ListenLimit)
 	}
 
 	if srv.TCPKeepAlive != 0 {

--- a/graceful.go
+++ b/graceful.go
@@ -437,29 +437,3 @@ func (srv *Server) shutdown(shutdown chan chan struct{}, kill chan struct{}) {
 	}
 	srv.chanLock.Unlock()
 }
-
-type keepAliveConn interface {
-	SetKeepAlive(bool) error
-	SetKeepAlivePeriod(d time.Duration) error
-}
-
-// keepAliveListener sets TCP keep-alive timeouts on accepted
-// connections. It's used by ListenAndServe and ListenAndServeTLS so
-// dead TCP connections (e.g. closing laptop mid-download) eventually
-// go away.
-type keepAliveListener struct {
-	net.Listener
-	keepAlivePeriod time.Duration
-}
-
-func (ln keepAliveListener) Accept() (net.Conn, error) {
-	c, err := ln.Listener.Accept()
-	if err != nil {
-		return nil, err
-	}
-
-	kac := c.(keepAliveConn)
-	kac.SetKeepAlive(true)
-	kac.SetKeepAlivePeriod(ln.keepAlivePeriod)
-	return c, nil
-}

--- a/graceful_test.go
+++ b/graceful_test.go
@@ -17,8 +17,11 @@ import (
 	"time"
 )
 
-// The tests will run a test server on this port.
-const port = 9654
+const (
+	// The tests will run a test server on this port.
+	port               = 9654
+	concurrentRequestN = 8
+)
 
 var (
 	killTime    = 500 * time.Millisecond
@@ -98,7 +101,7 @@ func launchTestQueries(t *testing.T, wg *sync.WaitGroup, c chan os.Signal) {
 	defer wg.Done()
 	var once sync.Once
 
-	for i := 0; i < 8; i++ {
+	for i := 0; i < concurrentRequestN; i++ {
 		wg.Add(1)
 		go runQuery(t, http.StatusOK, false, wg, &once)
 	}
@@ -107,7 +110,7 @@ func launchTestQueries(t *testing.T, wg *sync.WaitGroup, c chan os.Signal) {
 	c <- os.Interrupt
 	time.Sleep(waitTime)
 
-	for i := 0; i < 8; i++ {
+	for i := 0; i < concurrentRequestN; i++ {
 		wg.Add(1)
 		go runQuery(t, 0, true, wg, &once)
 	}
@@ -156,7 +159,7 @@ func TestGracefulRunTimesOut(t *testing.T) {
 		defer wg.Done()
 		var once sync.Once
 
-		for i := 0; i < 8; i++ {
+		for i := 0; i < concurrentRequestN; i++ {
 			wg.Add(1)
 			go runQuery(t, 0, true, &wg, &once)
 		}
@@ -165,7 +168,7 @@ func TestGracefulRunTimesOut(t *testing.T) {
 		c <- os.Interrupt
 		time.Sleep(waitTime)
 
-		for i := 0; i < 8; i++ {
+		for i := 0; i < concurrentRequestN; i++ {
 			wg.Add(1)
 			go runQuery(t, 0, true, &wg, &once)
 		}
@@ -226,9 +229,9 @@ func TestGracefulForwardsConnState(t *testing.T) {
 	defer wg.Wait()
 
 	expected := map[http.ConnState]int{
-		http.StateNew:    8,
-		http.StateActive: 8,
-		http.StateClosed: 8,
+		http.StateNew:    concurrentRequestN,
+		http.StateActive: concurrentRequestN,
+		http.StateClosed: concurrentRequestN,
 	}
 
 	c := make(chan os.Signal, 1)
@@ -393,7 +396,7 @@ func TestNotifyClosed(t *testing.T) {
 	}()
 
 	var once sync.Once
-	for i := 0; i < 8; i++ {
+	for i := 0; i < concurrentRequestN; i++ {
 		wg.Add(1)
 		runQuery(t, http.StatusOK, false, &wg, &once)
 	}

--- a/graceful_test.go
+++ b/graceful_test.go
@@ -137,6 +137,33 @@ func TestGracefulRun(t *testing.T) {
 	go launchTestQueries(t, &wg, c)
 }
 
+func TestGracefulRunLimitKeepAliveListener(t *testing.T) {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	c := make(chan os.Signal, 1)
+	server, l, err := createListener(killTime / 2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		srv := &Server{
+			Timeout:      killTime,
+			ListenLimit:  concurrentRequestN,
+			TCPKeepAlive: 1 * time.Second,
+			Server:       server,
+			interrupt:    c,
+		}
+		srv.Serve(l)
+	}()
+
+	wg.Add(1)
+	go launchTestQueries(t, &wg, c)
+}
+
 func TestGracefulRunTimesOut(t *testing.T) {
 	var wg sync.WaitGroup
 	defer wg.Wait()

--- a/graceful_test.go
+++ b/graceful_test.go
@@ -21,12 +21,9 @@ const (
 	// The tests will run a test server on this port.
 	port               = 9654
 	concurrentRequestN = 8
-)
-
-var (
-	killTime    = 500 * time.Millisecond
-	timeoutTime = 1000 * time.Millisecond
-	waitTime    = 100 * time.Millisecond
+	killTime           = 500 * time.Millisecond
+	timeoutTime        = 1000 * time.Millisecond
+	waitTime           = 100 * time.Millisecond
 )
 
 func runQuery(t *testing.T, expected int, shouldErr bool, wg *sync.WaitGroup, once *sync.Once) {

--- a/keepalive_listener.go
+++ b/keepalive_listener.go
@@ -1,0 +1,32 @@
+package graceful
+
+import (
+	"net"
+	"time"
+)
+
+type keepAliveConn interface {
+	SetKeepAlive(bool) error
+	SetKeepAlivePeriod(d time.Duration) error
+}
+
+// keepAliveListener sets TCP keep-alive timeouts on accepted
+// connections. It's used by ListenAndServe and ListenAndServeTLS so
+// dead TCP connections (e.g. closing laptop mid-download) eventually
+// go away.
+type keepAliveListener struct {
+	net.Listener
+	keepAlivePeriod time.Duration
+}
+
+func (ln keepAliveListener) Accept() (net.Conn, error) {
+	c, err := ln.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	kac := c.(keepAliveConn)
+	kac.SetKeepAlive(true)
+	kac.SetKeepAlivePeriod(ln.keepAlivePeriod)
+	return c, nil
+}

--- a/limit_listen.go
+++ b/limit_listen.go
@@ -15,9 +15,14 @@
 package graceful
 
 import (
+	"errors"
 	"net"
 	"sync"
+	"time"
 )
+
+// ErrNotTCP indicates that network connection is not a TCP connection.
+var ErrNotTCP = errors.New("only tcp connections have keepalive")
 
 // LimitListener returns a Listener that accepts at most n simultaneous
 // connections from the provided Listener.
@@ -53,4 +58,20 @@ func (l *limitListenerConn) Close() error {
 	err := l.Conn.Close()
 	l.releaseOnce.Do(l.release)
 	return err
+}
+
+func (l *limitListenerConn) SetKeepAlive(doKeepAlive bool) error {
+	tcpc, ok := l.Conn.(*net.TCPConn)
+	if !ok {
+		return ErrNotTCP
+	}
+	return tcpc.SetKeepAlive(doKeepAlive)
+}
+
+func (l *limitListenerConn) SetKeepAlivePeriod(d time.Duration) error {
+	tcpc, ok := l.Conn.(*net.TCPConn)
+	if !ok {
+		return ErrNotTCP
+	}
+	return tcpc.SetKeepAlivePeriod(d)
 }

--- a/limit_listen.go
+++ b/limit_listen.go
@@ -1,0 +1,56 @@
+// Copyright 2013 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graceful
+
+import (
+	"net"
+	"sync"
+)
+
+// LimitListener returns a Listener that accepts at most n simultaneous
+// connections from the provided Listener.
+func LimitListener(l net.Listener, n int) net.Listener {
+	return &limitListener{l, make(chan struct{}, n)}
+}
+
+type limitListener struct {
+	net.Listener
+	sem chan struct{}
+}
+
+func (l *limitListener) acquire() { l.sem <- struct{}{} }
+func (l *limitListener) release() { <-l.sem }
+
+func (l *limitListener) Accept() (net.Conn, error) {
+	l.acquire()
+	c, err := l.Listener.Accept()
+	if err != nil {
+		l.release()
+		return nil, err
+	}
+	return &limitListenerConn{Conn: c, release: l.release}, nil
+}
+
+type limitListenerConn struct {
+	net.Conn
+	releaseOnce sync.Once
+	release     func()
+}
+
+func (l *limitListenerConn) Close() error {
+	err := l.Conn.Close()
+	l.releaseOnce.Do(l.release)
+	return err
+}

